File: stroom-query-api/src/main/java/stroom/query/api/ExpressionOperator.java
/*
 * Copyright 2016 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stroom.query.api;

import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import stroom.util.shared.HasDisplayValue;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlElementWrapper;
import javax.xml.bind.annotation.XmlElements;
import javax.xml.bind.annotation.XmlType;
import java.util.List;

@JsonPropertyOrder({"op", "children"})
@XmlType(name = "ExpressionOperator", propOrder = {"op", "children"})
@XmlAccessorType(XmlAccessType.FIELD)
public final class ExpressionOperator extends ExpressionItem {
    private static final long serialVersionUID = 6602004424564268512L;

    @XmlElement(name = "op")
    private Op op = Op.AND;
    @XmlElementWrapper(name = "children")
    @XmlElements({
            @XmlElement(name = "operator", type = ExpressionOperator.class),
            @XmlElement(name = "term", type = ExpressionTerm.class)
    })
    private List<ExpressionItem> children;

    private ExpressionOperator() {
    }

    public ExpressionOperator(final Boolean enabled, final Op op, final List<ExpressionItem> children) {
        super(enabled);
        this.op = op;
        this.children = children;
    }

    public Op getOp() {
        return op;
    }

    public List<ExpressionItem> getChildren() {
        return children;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;

        final ExpressionOperator that = (ExpressionOperator) o;

        if (op != that.op) return false;
        return children != null ? children.equals(that.children) : that.children == null;
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + (op != null ? op.hashCode() : 0);
        result = 31 * result + (children != null ? children.hashCode() : 0);
        return result;
    }

    @Override
    public void append(final StringBuilder sb, final String pad, final boolean singleLine) {
        if (enabled()) {
            if (!singleLine && sb.length() > 0) {
                sb.append("\n");
                sb.append(pad);
            }

            sb.append(op);

            if (singleLine) {
                sb.append(" {");
            }

            if (children != null) {
                final String padding = pad + "  ";
                for (final ExpressionItem expressionItem : children) {
                    if (expressionItem.enabled()) {
                        if (!singleLine) {
                            sb.append("\n");
                            sb.append(pad);
                        }

                        expressionItem.append(sb, padding, singleLine);

                        if (singleLine) {
                            sb.append(", ");
                        }
                    }
                }

                if (singleLine) {
                    sb.setLength(sb.length() - ", ".length());
                }
            }

            if (singleLine) {
                sb.append("} ");
            }
        }
    }

    public enum Op implements HasDisplayValue {
        AND("AND"), OR("OR"), NOT("NOT");

        private final String displayValue;

        Op(final String displayValue) {
            this.displayValue = displayValue;
        }

        @Override
        public String getDisplayValue() {
            return displayValue;
        }
    }
}

File: stroom-ui/src/components/ExpressionBuilder/ExpressionOperator.js
/*
 * Copyright 2018 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { compose, withHandlers, withProps } from 'recompose';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from 'semantic-ui-react';
import { DragSource, DropTarget } from 'react-dnd';

import { canMove } from 'lib/treeUtils';
import ItemTypes from './dragDropTypes';
import ExpressionTerm from './ExpressionTerm';
import { actionCreators } from './redux';
import { LineTo } from 'components/LineTo';
import { LOGICAL_OPERATORS } from './logicalOperators';
import IconButton from 'components/IconButton';

const {
  expressionTermAdded,
  expressionOperatorAdded,
  expressionItemUpdated,
  expressionItemMoved,
  expressionItemDeleteRequested,
} = actionCreators;

const dragSource = {
  canDrag(props) {
    return true;
  },
  beginDrag(props) {
    return {
      ...props.operator,
    };
  },
};

function dragCollect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  };
}

const dropTarget = {
  canDrop(props, monitor) {
    return canMove(monitor.getItem(), props.operator);
  },
  drop(props, monitor) {
    props.expressionItemMoved(props.expressionId, monitor.getItem(), props.operator);
  },
};

function dropCollect(connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop(),
  };
}

const enhance = compose(
  connect(
    state => ({
      // operators are nested, so take all their props from parent
    }),
    {
      expressionTermAdded,
      expressionOperatorAdded,
      expressionItemUpdated,
      expressionItemMoved,
      expressionItemDeleteRequested,
    },
  ),
  DragSource(ItemTypes.OPERATOR, dragSource, dragCollect),
  DropTarget([ItemTypes.OPERATOR, ItemTypes.TERM], dropTarget, dropCollect),
  withHandlers({
    onAddOperator: ({ expressionOperatorAdded, expressionId, operator: { uuid } }) => {
      expressionOperatorAdded(expressionId, uuid);
    },

    onAddTerm: ({ expressionTermAdded, expressionId, operator: { uuid } }) => () => {
      expressionTermAdded(expressionId, uuid);
    },

    onOperatorUpdated: ({ expressionItemUpdated, expressionId, operator: { uuid } }) => (updates) => {
      expressionItemUpdated(expressionId, uuid, updates);
    },

    onOpChange: ({ expressionItemUpdated, expressionId, operator: { uuid } }) => (op) => {
      expressionItemUpdated(expressionId, uuid, {
        op,
      });
    },

    onRequestDeleteOperator: ({
      expressionItemDeleteRequested,
      expressionId,
      operator: { uuid },
    }) => () => {
      expressionItemDeleteRequested(expressionId, uuid);
    },

    onEnabledToggled: ({
      isRoot,
      expressionItemUpdated,
      expressionId,
      operator: { uuid, enabled },
    }) => () => {
      if (!isRoot) {
        expressionItemUpdated(expressionId, uuid, {
          enabled: !enabled,
        });
      }
    },
  }),
  withProps(({
    canDrop, isOver, isRoot, isEnabled, operator,
  }) => {
    let dndBarColour = 'grey';
    if (isOver) {
      dndBarColour = canDrop ? 'blue' : 'red';
    }

    const classNames = ['expression-item'];
    if (isRoot) {
      classNames.push('expression-item__root');
    }
    if (!isEnabled) {
      classNames.push('expression-item--disabled');
    }

    let enabledIcon = 'check';
    let enabledColour = 'grey';
    if (isRoot) {
      enabledIcon = 'ban';
    } else if (operator.enabled) {
      enabledColour = 'blue';
    }

    return {
      enabledIcon,
      enabledColour,
      dndBarColour,
      className: classNames.join(' '),
    };
  }),
);

const ExpressionOperator = ({
  expressionId,
  operator,
  isRoot,
  isEnabled,
  dataSource,

  connectDropTarget,
  isOver,
  connectDragSource,
  isDragging,

  dndBarColour,
  className,

  onAddOperator,
  onAddTerm,
  onOperatorUpdated,
  onOpChange,
  onRequestDeleteOperator,
  onEnabledToggled,

  enabledIcon,
  enabledColour,
}) => (
    <div className={className}>
      {connectDropTarget(<div>
        {connectDragSource(<span id={`expression-item${operator.uuid}`}>
          <FontAwesomeIcon color={dndBarColour} icon="bars" />
        </span>)}

        <Button.Group>
          {LOGICAL_OPERATORS.map(l => (
            <Button
              color={operator.op === l ? 'blue' : undefined}
              key={l}
              compact
              onClick={() => onOpChange(l)}
            >
              {l}
            </Button>
          ))}
        </Button.Group>

        <IconButton icon="plus" text="Term" groupPosition='left' onClick={onAddTerm} />
        <IconButton icon="plus" text="Group" groupPosition='middle' onClick={onAddOperator} />
        <IconButton icon={enabledIcon} groupPosition='middle' color={enabledColour} onClick={onEnabledToggled} />
        {!isRoot ?
          (<IconButton icon="trash" groupPosition='right' onClick={onRequestDeleteOperator} />)
          :
          (<IconButton disabled icon="ban" groupPosition='right' />)
        }

      </div>)}

      <div className="operator__children">
        {isOver && dropTarget.canDrop && <div className="operator__placeholder" />}
        {operator.children
          .map((c) => {
            let itemElement;
            switch (c.type) {
              case 'term':
                itemElement = (
                  <div key={c.uuid} id={`expression-item${c.uuid}`}>
                    <ExpressionTerm
                      dataSource={dataSource}
                      expressionId={expressionId}
                      isEnabled={isEnabled && c.enabled}
                      term={c}
                    />
                  </div>
                );
                break;
              case 'operator':
                itemElement = (
                  <EnhancedExpressionOperator
                    dataSource={dataSource}
                    expressionId={expressionId}
                    isEnabled={isEnabled && c.enabled}
                    operator={c}
                  />
                );
                break;
              default:
                throw new Error(`Invalid operator type: ${c.type}`);
            }

            // Wrap it with a line to
            return (
              <div key={c.uuid}>
                <LineTo
                  lineId={c.uuid}
                  lineType="downRightElbow"
                  fromId={`expression-item${operator.uuid}`}
                  toId={`expression-item${c.uuid}`}
                />
                {itemElement}
              </div>
            );
          })
          .filter(c => !!c) // null filter
        }
      </div>
    </div>
  );

const EnhancedExpressionOperator = enhance(ExpressionOperator);

EnhancedExpressionOperator.propTypes = {
  dataSource: PropTypes.object.isRequired, // complete definition of the data source
  expressionId: PropTypes.string.isRequired, // the ID of the overall expression
  operator: PropTypes.object.isRequired, // the operator that this particular element is to represent
  isRoot: PropTypes.bool.isRequired, // used to prevent deletion of root nodes
  isEnabled: PropTypes.bool.isRequired, // a combination of any parent enabled state, and its own
};

EnhancedExpressionOperator.defaultProps = {
  isRoot: false,
};

export default EnhancedExpressionOperator;
